lwip/cannelloni.c

```diff
@@ -57,112 +96,117 @@ void handle_cannelloni_frame(void *arg, struct udp_pcb *pcb, struct pbuf *p, con
             error = 1;
             break;
           }
-          memcpy(frame->data, rawData, canfd_len(frame));
-          rawData += canfd_len(frame);
+        }
+
+        struct canfd_frame *frame = queue_put(&handle->tx_queue);
+        if (!frame) {
+          /* Allocation error */
+          error = 1;
+          break;
+        }
+        frame->can_id = can_id;
+        frame->len = len;
+        frame->flags = flags;
+
+        /* RTR Frames have no data */
+        if ((can_id & CAN_RTR_FLAG) == 0) {
+          memcpy(frame->data, rawData + pos, dlc);
+          pos += dlc;
         }
       }
     }
+  }
+
+  if(p) {
     pbuf_free(p);
   }
 }
 
 void transmit_udp_frame(cannelloni_handle_t* handle)
 {
-  if (handle->can_rx_frames_pos >= 0) {
-    struct cannelloni_data_packet *dataPacket;
-
-    struct pbuf *p = pbuf_alloc(PBUF_TRANSPORT, 1200, PBUF_RAM);
-    uint16_t length = 0;
-    uint16_t frameCount = 0;
-    struct canfd_frame *frame;
-    uint8_t *data = (uint8_t*) p->payload + CANNELLONI_DATA_PACKET_BASE_SIZE;
-    if (!p) {
-      /* allocation error */
-      return;
-    }
-
-    while(handle->can_rx_frames_pos >= 0) {
-      frame = &handle->Init.can_rx_buf[handle->can_rx_frames_pos];
-      data[0] = frame->can_id >> 24 & 0x000000ff;
-      data[1] = frame->can_id >> 16 & 0x000000ff;
-      data[2] = frame->can_id >> 8  & 0x000000ff;
-      data[3] = frame->can_id       & 0x000000ff;
-      /* += 4 */
-      data += sizeof(canid_t);
-      *data = frame->len;
-      /* += 1 */
-      data += sizeof(frame->len);
-      memcpy(data, frame->data, canfd_len(frame));
-      data+=canfd_len(frame);
-      frameCount++;
-      handle->can_rx_frames_pos--;
-    }
-
-    dataPacket = (struct cannelloni_data_packet*) p->payload;
-    dataPacket->version = CANNELLONI_FRAME_VERSION;
-    dataPacket->op_code = CNL_DATA;
-    dataPacket->seq_no = handle->sequence_number++;
-    dataPacket->count = htons(frameCount);
-
-    length = (uint8_t*)data-(uint8_t*)p->payload;
-    p->tot_len = length;
-    p->len = length;
+  struct canfd_frame *frame = queue_take(&handle->rx_queue);
+  if (!frame) {
+    return;
+  }
 
-    udp_sendto(handle->udp_pcb, p, &(handle->Init.addr), handle->Init.remote_port);
-    pbuf_free(p);
+  struct pbuf *p = pbuf_alloc(PBUF_TRANSPORT, 1200, PBUF_RAM);
+  uint16_t length = 0;
+  uint16_t frameCount = 0;
+  uint8_t *data = (uint8_t*) p->payload + CANNELLONI_DATA_PACKET_BASE_SIZE;
+  if (!p) {
+    /* allocation error */
+    return;
   }
+
+  do {
+    data[0] = frame->can_id >> 24 & 0x000000ff;
+    data[1] = frame->can_id >> 16 & 0x000000ff;
+    data[2] = frame->can_id >> 8  & 0x000000ff;
+    data[3] = frame->can_id       & 0x000000ff;
+    /* += 4 */
+    data += sizeof(canid_t);
+    *data = frame->len;
+    /* += 1 */
+    data += sizeof(frame->len);
+    memcpy(data, frame->data, canfd_len(frame));
+    data+=canfd_len(frame);
+    frameCount++;
+  } while (frame = queue_take(&handle->rx_queue));

```

Comment:
Not introduced by you, but probably a good idea to fix when touching this part:

If this function is not called often enough, the resulting `dataPacket` might be larger than the MTU.
There should be another condition that leaves this queue, i.e. not sending the whole queue.
I suggest to add a simply return code indicating whether `transmit_udp_frame` should be called again to drain the queue.

Reply:
`transmit_udp_frame` is now called repeatedly when there are more frames in queue and `pbuf` allocation was successful.

Did the same thing also for the other queue - we are putting CAN frames to CAN driver (`can_tx_fn`) until it stops processing them (because mailboxes are full etc)